app/Controller.cpp

```diff
@@ -80,9 +85,16 @@ void Controller::controlLoop() {
     // update next target time
     next_loop_time += duration;
 
-    //
-    // @TODO groundbreaking robotics goes here
-    //
+    double speed_error, heading_error, command_throttle, command_steering, actual_throttle, actual_steering;
+    double steeringVel = 0.0; // TODO: Understanding the current_steering_vel and update accordingly
+
+    this->model_->getError(speed_error, heading_error);
+    command_throttle = this->pid_speed_->getCommand(speed_error);
+    command_steering = this->pid_heading_->getCommand(heading_error);
+  
+    this->model_->limits_->limit(command_throttle, command_steering, steeringVel, actual_throttle, actual_steering, steeringVel, duration);
+  
+    this->model_->command(actual_throttle, actual_steering, duration);
```

Comment:
This is the wrong duration variable; you'll need to convert it to a double to get this to run.